### ISSUE TITLE
Enumeration generates invalid dimensions at observation nodes

### ISSUE THREAD

**Issue by fehiepsi** [pyro-ppl/numpyro#779]
Copied from #778

Consider the model (corrected thanks to @eb8680 )
```python
def model(X, y):
    N, d = X.shape
    with numpyro.plate('d', d, dim=-1):
        pi = numpyro.sample('pi', dist.Beta(0.5, 0.5))
        w = numpyro.sample('w', dist.Normal(0., 10.0))

    ws = []
    for k in range(d):
        xi_k = numpyro.sample(f'xi_{k}', dist.Bernoulli(probs=pi[k]))
        w_k = np.where(xi_k, w[k:k + 1], 0) 
        ws.append(w_k)
    ws = np.concatenate(np.broadcast_arrays(*ws), axis=-1)

    loc = np.einsum("...nd,...d->...n", X, ws)
    with numpyro.plate('N', N, dim=-1):
        return numpyro.sample('obs', dist.Normal(loc), obs=y)
```

<details> <summary>Full code</summary>

```python
import jax
from jax import random
import jax.numpy as np
import numpyro
import numpyro.distributions as dist
from numpyro.contrib.indexing import Vindex
from numpyro.infer import NUTS, MCMC
jax.config.disable_omnistaging()

def generate_data(key, n_samples=100, n_features=100, features_kept=10):
    X = random.normal(key, (n_samples, n_features))
    # Create weights with a precision lambda_ of 4.
    lambda_ = 4.
    w = np.zeros(n_features)
    # Only keep 10 weights of interest
    relevant_features = random.choice(
        key, np.arange(n_features), (features_kept,), False)
    w = jax.ops.index_update(
        w,
        relevant_features,
        dist.Normal(loc=0, scale=1. / np.sqrt(lambda_)).sample(key, (features_kept,))
    )
    # Create noise with a precision alpha of 50.
    alpha_ = 50.
    noise = dist.Normal(loc=0, scale=1. / np.sqrt(alpha_)).sample(key, (n_samples,))
    # Create the target
    y = np.dot(X, w) + noise
    return X, y, w

X, y, w = generate_data(random.PRNGKey(0), 4, 3, 2)
X.shape, y.shape

mcmc = MCMC(NUTS(model), 500, 500)
mcmc.run(random.PRNGKey(0), X, y)
```
</details>

The factors will be
```
[Tensor(Traced<ConcreteArray([-1.401806  -2.1019688 -1.7583057])>with<JVPTrace(level=1/0)>
  with primal = DeviceArray([-1.401806 , -2.1019688, -1.7583057], dtype=float32)
       tangent = Traced<ShapedArray(float32[3]):JaxprTrace(level=0/0)>, OrderedDict([('d', Bint[3, ])]), 'real'), Tensor(Traced<ConcreteArray([-0.44382608 -0.09374487 -0.2655763 ])>with<JVPTrace(level=1/0)>
  with primal = DeviceArray([-0.44382608, -0.09374487, -0.2655763 ], dtype=float32)
       tangent = Traced<ShapedArray(float32[3]):JaxprTrace(level=0/0)>, OrderedDict([('d', Bint[3, ])]), 'real'), Tensor(Traced<ConcreteArray([-3.229965  -3.2219462 -3.239487 ])>with<JVPTrace(level=1/0)>
  with primal = DeviceArray([-3.229965 , -3.2219462, -3.239487 ], dtype=float32)
       tangent = Traced<ShapedArray(float32[3]):JaxprTrace(level=0/0)>, OrderedDict([('d', Bint[3, ])]), 'real'), Tensor(Traced<ConcreteArray([-0.82561004 -0.576196  ])>with<JVPTrace(level=1/0)>
  with primal = DeviceArray([-0.82561004, -0.576196  ], dtype=float32)
       tangent = Traced<ShapedArray(float32[2]):JaxprTrace(level=0/0)>, OrderedDict([('xi_0', Bint[2, ])]), 'real'), Tensor(Traced<ConcreteArray([-1.9481988  -0.15376975])>with<JVPTrace(level=1/0)>
  with primal = DeviceArray([-1.9481988 , -0.15376975], dtype=float32)
       tangent = Traced<ShapedArray(float32[2]):JaxprTrace(level=0/0)>, OrderedDict([('xi_1', Bint[2, ])]), 'real'), Tensor(Traced<ConcreteArray([-0.25015315 -1.5081525 ])>with<JVPTrace(level=1/0)>
  with primal = DeviceArray([-0.25015315, -1.5081525 ], dtype=float32)
       tangent = Traced<ShapedArray(float32[2]):JaxprTrace(level=0/0)>, OrderedDict([('xi_2', Bint[2, ])]), 'real'), Tensor(Traced<ConcreteArray([[[[-1.3046664  -0.977973   -1.213804   -0.9663509 ]
   [-1.1421189  -1.0131184  -0.93338925 -2.189458  ]]

  [[-1.6366069  -1.0236161  -1.8565254  -1.0466189 ]
   [-0.9796202  -0.9701585  -1.2164943  -2.5236025 ]]]


 [[[-0.91963637 -1.3886205  -1.6084142  -0.9411368 ]
   [-2.0575821  -2.4446356  -1.0850296  -1.4972053 ]]

  [[-0.982695   -1.2846785  -2.4955258  -0.9190271 ]
   [-1.6262016  -2.25209    -1.6125246  -1.7289721 ]]]])>with<JVPTrace(level=1/0)>
  with primal = DeviceArray([[[[-1.3046664 , -0.977973  , -1.213804  , -0.9663509 ],
                               [-1.1421189 , -1.0131184 , -0.93338925, -2.189458  ]],
                
                              [[-1.6366069 , -1.0236161 , -1.8565254 , -1.0466189 ],
                               [-0.9796202 , -0.9701585 , -1.2164943 , -2.5236025 ]]],
                
                
                             [[[-0.91963637, -1.3886205 , -1.6084142 , -0.9411368 ],
                               [-2.0575821 , -2.4446356 , -1.0850296 , -1.4972053 ]],
                
                              [[-0.982695  , -1.2846785 , -2.4955258 , -0.9190271 ],
                               [-1.6262016 , -2.25209   , -1.6125246 , -1.7289721 ]]]],            dtype=float32)
       tangent = Traced<ShapedArray(float32[2,2,2,4]):JaxprTrace(level=0/0)>, OrderedDict([('_pyro_dim_5', Bint[2, ]), ('xi_2', Bint[2, ]), ('xi_1', Bint[2, ]), ('d', Bint[4, ])]), 'real')]
```

We can see that the factor `obs` has wrong dimensions `_pyro_dim_5`, `xi_2`, `xi_1`, `d`. I would expect the correct dimensions are `xi_2`, `xi_1`, `xi_0`, and `N`.

**Comment by eb8680** [comment]
I probably won't have time until after PROBPROG to look into this (I'll update `numpyro.contrib.funsor` as part of #741), but unfortunately it's not just a bug - the code as written is underspecified.  In particular, dimension `-1` means different things at different points in the model, and values carrying those different meanings interact in the final lines.

A less ambiguous version of this code would not use a plate at the beginning and would avoid `matmul` in favor of `einsum`:
```py
def model(X, y):
    N, d = X.shape
    assert y.shape == (N,)
    pi = numpyro.sample('pi', dist.Beta(0.5, 0.5).expand([d]).to_event(1))
    w = numpyro.sample('w', dist.Normal(0., 10.0).expand([d]).to_event(1))

    ws = []
    for k in range(d):
        xi_k = numpyro.sample(f'xi_{k}', dist.Bernoulli(probs=pi[k]))
        w_k = np.where(xi_k, w[k], 0)
        ws.append(w_k)
    ws = np.stack(np.broadcast_arrays(*ws), axis=-1)

    loc = np.einsum("...nd,...d->...n", X, ws)
    with numpyro.plate('N', N, dim=-1):
        return numpyro.sample('obs', dist.Normal(loc, 1.), obs=y)
```

**Comment by fehiepsi** [comment]
Thanks! I intended to look at this later. Your code still generates incorrect named dimensions `OrderedDict([('_pyro_dim_5', Bint[2, ]), ('xi_2', Bint[2, ]), ('xi_1', Bint[2, ]), ('N', Bint[4, ])]), 'real')`. I think it is broken at some points after your original PR (HMM example `model4` is broken).

**Comment by eb8680** [comment]
I see, seems like a subtle bug snuck into `plate` somewhere after we forgot to add the HMM example to `test_examples`. Sorry about that!

**Comment by fehiepsi** [comment]
You are right. We need to add HMM to `test_examples`. I think the bug will be simple to detect so I'll walk through the history first.

**Comment by fehiepsi** [comment]
@eb8680 It turns out that my code to compute `ws` is incorrect. HMM failed when I tried `hidden_dim=2`, which even failed for our earliest support of `markov`. But it passes for larger hidden dims. So far, the only regression is at `plate.postprocess_msg`, which will also be addressed in #780. I will explain the situation in more details in that PR.
